Bringing Matrix to Elixir post

niklaslong opened the pull request:
There a still a few things I'd like to tweak over the weekend but I thought I'd open the PR if anyone is curious.

marcoow commented on _posts/2020-08-22-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,194 @@
+{:ok,
+ %Tesla.Env{
+   body: %{
+     "consent_uri" => "https://matrix-client.matrix.org/_matrix/consent?",
```
Would this have to include some kind of token so it's able to connect the consent to this particular session?

niklaslong replied:
Yes, I shortened the url but could add it back in or make a comment of the removal?

marcoow replied:
I'd maybe leave a bit of the token in at least since otherwise this is confusing

marcoow commented on _posts/2020-08-22-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,194 @@
+
+This is where I'll end this short introduction, however, please check the
+[documentation](https://hexdocs.pm/matrix_sdk/) for more information on the
+currently implemented endpoints.
```
Would it be interesting to add some insights into how you built this with Elixir and why Elixir is a particularly good fit for implementing such a protocol maybe?

niklaslong replied:
I make a brief mention of this in the conclusion and I've added a note about the SDK's modularity. The implementation details aren't that interesting yet, as it's essentially an API wrapper at this point.

marcoow reviewed (commented):
Nice post! I left some comments.

Didn't know about Matrix…

niklaslong commented on _posts/2020-08-22-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,194 @@
+## Down the rabbit-hole...
+
+Matrix is structured around the federation of homeservers, that is to say the
+continuous synchronisation of event history between the homeservers via the
```
Yes, it is 👍

niklaslong commented on _posts/2020-08-22-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,194 @@
+A note on Matrix interoperability: it was designed to exchange data with other
+platforms such as WhatsApp, Slack, iMessage, Email, Discord, IRC and many more.
+This is known as bridging and makes Matrix an attractive one-stop solution to
+interface with these services.
```
I've added a few details, though actually going through an example is beyond the scope of this post as it deals with an area of Matrix I'm not directly working on. There are also many different ways to bridge, the details of which are currently highly service dependent (this is integration territory).

marhigh commented:
Nice sketch 👌

niklaslong commented:
👋 @marcoow What's the status on this? 🙏

marcoow commented:
@niklaslong sorry, was super busy with other things in the past days – I should get to this tomorrow 🙏

niklaslong commented:
No worries, thanks @marcoow ! 🙌

marhigh commented:
Sure! Any requests on what the visual should contain?

On Fri, 9 Oct 2020 at 18:05, Marco Otte-Witte <notifications@github.com>
wrote:

> *@marcoow* commented on this pull request.
>
> Great post! @marhigh <https://github.com/marhigh> we need an illustration
> and og:image – can you add those?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/simplabs/simplabs.github.io/pull/1257#pullrequestreview-505808211>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB6YKQGZ7CR33GIFKNDAWV3SJ4YDBANCNFSM4QOBENDA>
> .
>


-- 
Mar High
Director of Product Design
https://simplabs.com

simplabs GmbH
Hans-Sachs-Str. 12
80469 München
Managing Director: Marco Otte-Witte
Handelsgericht München HRB 217887

marcoow commented:
@niklaslong can you figure out a suitable illustration with @marhigh ?

niklaslong commented:
@marhigh as this is essentially about decentralised nature of matrix based communication, I think it would be good to make that clear in the illustration (the typical diagram to describe matrix's architecture succinctly is about a third of the way down on their homepage: https://matrix.org/), I also quite like the symbolism of network graphs (as they are quite beautiful and easily understood?). Elixir being used for this should probably also be subtly hinted at (purple being used for the _connections_ perhaps)? This is also a library meant for developers, perhaps this should be hinted at in some way...

These are just my initial thoughts and would be happy to brainstorm further if these are too abstract :) 
Please let me know if there's a specific type of input which would be helpful?

marhigh commented:
@niklaslong that is a great start! I'll create a visual direction and share it with you via abstract in a little bit, then we can continue any feedback over there.

marhigh commented:
@niklaslong I just emailed you an invite to abstract to niklas@long.ch. Once accepted, I'll add you as the reviewer of this branch so you can make any comments for changes or download the asset (svg). I also added an og:image to this.

niklaslong commented:
I've added the og:image and illustration (after digging around on abstract could only find the `svg` for this asset, though I think it's meant to be a `png` as it's not being rendered as with other posts?)

Note: we should also remember to change the post date when publishing as it's quite outdated.

marhigh commented:
@niklaslong the illustration is meant to be svg, wheras the og:image is a png. 

We haven't used svgs in the past, but we should be switching to them now for this type of asset. They are much smaller in file size and maintain image integrity in larger viewports. 

Please refer to Ricardo's latest post if you need a hint on how to use it: https://simplabs.com/blog/2020/10/05/simpler-and-more-powerful-components-in-ember-octane-with-glimmer-components/

niklaslong commented:
Fixed—thanks @marhigh, somehow I missed it being used in the last post 👍

marcoow commented:
@niklaslong Nice – thanks for this! I asked @Kristina2007 (our marketing student) to coordinate with you regarding the release (she'll review for SEO potential again and prepare Tweets, etc. for which she will need a bit of your input)

marcoow commented:
@niklaslong ah, one thing – before publishing, please make sure to update the post date to the actual publishing date (once Kristina and you will know when that is) which means changing the file name as well as the image's folder name 👍

niklaslong commented:
Note: converting to draft as some changes will need to be made to the examples following [this PR](https://github.com/niklaslong/matrix-elixir-sdk/pull/83) which significantly changes the SDK's API.

Kristina2007 commented:
Please ignore me marking the pull request as ready for review.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+---
```
Sorry for the delay but will add all my comments now. I can also try to do the changes tommorow myself if you guys want to apply some of them.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+---
```
Sorry for the delay but will add all my comments now. I can also try to do the changes tommorow myself if you guys want to apply some of them.

Kristina2007 replied:
General points: 
-> Spaces between headlines are very big
-> Headlines seem disproportionately big in comparison to the rest of the text

niklaslong replied:
Formatting is determined by the design of the website. I've used h3s in place of h2s for this second draft, though I'm not sure what is currently preferred?

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+---
+title: Bringing the Matrix Open Standard to Elixir
```
From SEO perspective a formulation like this would be better: How to bring the Matrix Open Standard to Elixir (for example) It should be ideally formulated like a question or answer.

niklaslong replied:
That would be difficult in this case given that the post is introducing a tool and not really explaining _how_ to do something. I'll keep it in mind for future posts though, thanks for the advice 👍

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+open standard communication protocol designed to protect people's privacy. The
+technology has applications not only in messaging and VoIP but similarly in IoT
+and AR/VR.
+
```
From SEO perspective it makes sense spell out abbreviations in the beginning once and then use the abbreviations throughout the rest of the text. (create long tail keywords/ readability/ explain words)

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+[Matrix SDK for Elixir](https://github.com/niklaslong/matrix-elixir-sdk) with
+the aim of simplifying the process of Matrix-enabling Elixir applications. It's
+early days for the project but if you are interested in contributing (all skill
+levels welcome) or using the SDK as a foundation for another project, please let
```
1) Also would be better to spell out SDK once
2) using SDK as foundation for another project -> would be good to specify what kind of projects (explanatory/ long tail keyword) -> other Elixir projects?

marcoow replied:
Nobody spells out SDK in tech ;) I understand spelling out other things makes sense but not for this one.

Kristina2007 replied:
Okay, I understand. I was wondering about that too. Not everything beneficiary for SEO will necessarily be making sense for Tech topics.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+
+## Down the rabbit-hole...
+
+Matrix is structured around the federation of homeservers, that is to say the
```
Suggestion: Matrix is structured around the federation of homeservers which can be understood as the continuous synchronisation of event history between the homeservers via the Server API.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+A note on Matrix interoperability: it was designed to exchange data with other
+platforms such as WhatsApp, Slack, iMessage, Email, Discord, IRC and many more.
+This is known as bridging and makes Matrix an attractive one-stop solution to
+interface with these services. As an example, Alice who's on Matrix, could
```
What kind of services? Would be better to explain more specifically. 
-> communication services?

niklaslong replied:
I think the type of services is sufficiently implied by the examples and the nature of Matrix?

Kristina2007 replied:
It definitely is. Just from an SEO perspective it makes sense to specify such terms but it was just an idea. But this version also works if you prefer to keep it that way.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+## ...and into the Matrix
+
+The Elixir SDK currently wraps part of the Client-Server API; v0.1 was released
+to allow interested parties to begin experimenting. The `Client` module, used to
```
The Elixir SDK currently wraps part of the Client-Server API; v0.1 was released to allow interested parties to begin experimenting. 
-> Is there possibly a way to make this a fully connected sentence to improve the readability?

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+
+The `Client.register_guest/1` call registers a guest user with the Matrix
+homeserver and returns an access token (I have shortened for brevity). This
+token can be used to authenticate most Matrix endpoints: some don't require
```
Suggestion: This token can be used to authenticate most Matrix endpoints while on the contrary some don't require authentication at all.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+The Elixir SDK currently wraps part of the Client-Server API; v0.1 was released
+to allow interested parties to begin experimenting. The `Client` module, used to
+make requests to homeservers, is currently the highest level of abstraction in
+the SDK. It is designed to be modular; this, for example, makes it easy to
```
Suggestion: It is designed to be modular which makes it easy to switch out the HTTP layer if you choose to do so.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+
+As mentioned above, all changes in a room's state are described by events. They
+can represent any data, from users joining a room or sending messages, to image
+uploads and VoIP call setup. Let's dip our toes into Matrix by creating a guest
```
->setups?
-> Suggestion: So let's dip our toes into the Matrix by creating a guest account on matrix.org......

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
````diff
@@ -0,0 +1,204 @@
+```
+
+Once we have an access token, we can attempt to join a room allowing guest
+access. One such room is `#matrix:matrix.org`.
````
Suggestion: This room is called #matrix:matrix.org./ One such room ist called #matrix:matrix.org.

niklaslong commented:
@Kristina2007, thanks for the comments! 

I won't have time to rewrite parts of the post to reflect the changes to the SDK before this weekend. In the spirit of avoiding a rushed job, as well as leaving some time for potential edits, it makes sense to me to push the publishing of the post to Nov. 2 (you mentioned Mondays are better)?

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+
+The call returns a `200` and the `room_id`. You may have noticed this isn't the
+same as `#matrix:matrix.org` used to make the request. The latter is an alias.
+They allows users to refer to rooms more conveniently instead of using long IDs
```
They allows?
Suggestion: They allow/ It allows

niklaslong replied:
Good catch! 👍

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
````diff
@@ -0,0 +1,204 @@
+```
+
+Syncing is complex and I won't be going into any great detail here. At a high
+level, all events from a user's joined rooms will be included in the response.
````
Suggestion: At a high level, all events of the rooms a user joined, will be included in the response.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+
+## What's next?
+
+Work continues on the SDK, there are a number of endpoints waiting to be
```
Suggestion: As work continues on the SDK, there are a number of endpoints waiting to be implemented and decisions to be made in how to handle state.

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+could lead to interesting implementations in Elixir, potentially targeting
+WebAssembly thanks to [Lumen](https://github.com/lumen/lumen).
+
+It is my belief Elixir can be a powerful tool in decentralising the web.
```
Suggestion: It is my belief that/ It is my belief, Elixir can be a powerful tool in decentralising the web. Projects like Matrix are vitally important and therefore I hope the SDK will encourage creators to start projects in this problem space.

niklaslong replied:
Switched out the comma for an _and_ 👍

niklaslong commented:
@Kristina2007, thanks for the comments! 

I won't have time to rewrite parts of the post to reflect the changes to the SDK before this weekend. In the spirit of avoiding a rushed job, as well as leaving some time for potential edits, it makes sense to me to push the publishing of the post to Nov. 2 (you mentioned Mondays are better)?

Kristina2007 commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,204 @@
+  image: /assets/images/posts/2020-10-26-bringing-matrix-to-elixir/og-image.png
+---
+
+[Matrix](https://matrix.org) is an open-source, end-to-end encrypted, real-time,
```
Nice!

Kristina2007 replied:
Nice explanation

Kristina2007 commented:
That would work for me if that is fine for everyone else.

niklaslong commented:
@marcoow, @Kristina2007, I've made some changes to the example section and reworked some parts (including Kristina's feedback—thank you 🙌) to reflect the changes to the SDK since the first draft and should now be up to date. I've also changed the title to something slightly simpler. 

What header sizes are we using, Kristina mentioned the h2s were big and this draft now uses h3s and h4s?

marcoow commented:
@niklaslong we should be using h2 and below but in the correct hierarchy

marcoow commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,276 @@
+[an example script](https://github.com/niklaslong/matrix-elixir-sdk/blob/master/examples/guest_login.exs)
+to do this and will be going through the crucial parts below. If you'd like to
+try it out yourself, clone the repo and run `mix run examples/guest_login.exs`.
+Note: you may need to run `mix deps.get` as well.
```
```suggestion
Note: you will need to run `mix deps.get` as well.
```

?

marcoow commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,276 @@
+  "sender" => "@56440647111:matrix.org",
+  "state_key" => "@56440647111:matrix.org",
+  "type" => "m.room.member",
+    "unsigned" => %{"age" => 5068277}
```
```suggestion
  "unsigned" => %{"age" => 5068277}
```

marcoow commented on _posts/2020-10-26-bringing-matrix-to-elixir.md:
```diff
@@ -0,0 +1,276 @@
+[documentation](https://hexdocs.pm/matrix_sdk/) for more information on the
+currently implemented endpoints.
+
+### Encryption with Olm
```
The jump from the previous paragraph to this is a bit direct – maybe there's a way to connect the 2 better?

niklaslong replied:
Agreed, I've moved it into the _What's next_ section 👍

marcoow commented:
Nice! Just one thing – the `og:image` still has the old title – @marhigh, can you update that? I couldn't find the `og:image` for this post on Abstract…

niklaslong commented:
Ah, good catch! 🙌

marhigh commented:
Og:image updated 👍

marcoow commented:
@niklaslong sorry, one last thing – can you update the date (filename and folder name for the images) to match the publication date, Nov 2nd?
